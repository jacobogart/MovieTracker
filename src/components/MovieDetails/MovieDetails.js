import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { CastContainer } from '../CastContainer/CastContainer';
import { fetchAddFavorite } from '../../api/fetchAddFavorite';

class MovieDetails extends Component {
  state = {
    error: ''
  }

  handleFavorite = () => {
    const { user, movieDetails } = this.props;

    if (user.id && !user.favorites.includes(movieDetails.id)) {
      fetchAddFavorite({ ...movieDetails,  user })
        .catch(error => this.setState({ error }));
      this.props.updateUserFavorites(movieDetails.id);
    } else if (user.id && user.favorites.includes(movieDetails.id)){
      this.props.deleteUserFavorite(user.id, movieDetails.id);
    } else {
      alert('You must be logged in to favorite a movie')
    }
  }

  render() {
    const { props } = this;
    const { title, overview, posterImg, id, rating, releaseDate, backdropImg } = props.movieDetails;
    const { director, writer, cast } = props.people;
    const backgroundImg = {
      backgroundImage: `url(${backdropImg})`
    }
    let releaseYear = releaseDate.substring(0, 4);
    let percentage = rating * 10;
    let ratingNum;
    let directorInfo;
    let writerInfo;
    let heartClass = 'far';

    if (props.user.id && props.user.favorites.includes(id)) {
      heartClass = 'fas';
    }

    if (rating.toString().length === 1) {
      ratingNum = `${rating}.0`;
    } else {
      ratingNum = rating;
    };

    let ratingCircle = 
      <svg viewBox="0 0 36 36" className="rating-circle">
        <path className="circle-bg"
          d="M18 2.0845
            a 15.9155 15.9155 0 0 1 0 31.831
            a 15.9155 15.9155 0 0 1 0 -31.831"/>
        <path className="circle"
          strokeDasharray={`${percentage}, 100`}
          d="M18 2.0845
            a 15.9155 15.9155 0 0 1 0 31.831
            a 15.9155 15.9155 0 0 1 0 -31.831"/>
        <text x="9" y="23" className="rating-num">{ratingNum}</text>
      </svg>

    if (director) {
      directorInfo = director.map(person => {
        return (
          <div className="crew-member">
            <h4>{person.job}</h4>
            <p>{person.name}</p>
          </div>
        );
      });
    }

    if (writer) {
      writerInfo = writer.map(person => {
        return (
          <div className="crew-member">
            <h4>{person.job}</h4>
            <p>{person.name}</p>
          </div>
        );
      });
    }

    return (
      <div className="MovieDetails-container">
        <div className="MovieDetails-backdrop" style={backgroundImg}>
          <div className="backdrop-cover">
            <article className="MovieDetails">
              <img className="card-img" src={posterImg} alt={`${title} poster`} />
              <div className="card-info">
                <div className="card-header">
                  <div className="card-title-wrap">
                    <h2 className="card-title">{title}</h2>
                    <p className="card-year">({releaseYear})</p>
                  </div>
                  <div className="card-btns">
                    <div className="circle-wrapper">{ratingCircle}</div>
                    <div 
                      className="btn-wrapper heart-wrapper"
                      onClick={this.handleFavorite}>
                      <i className={`${heartClass} fa-heart`} />
                    </div>
                    <div className="btn-wrapper bookmark-wrapper"><i class="far fa-bookmark"></i></div>
                  </div>
                </div>
                <p className="card-overview">{overview}</p>
                <div className="crew-info">
                  {directorInfo}
                  {writerInfo}
                </div>
              </div>
              <p className="card-overview">{overview}</p>
              <div className="crew-info">
                {directorInfo}
                {writerInfo}
              <div className="card-btns">
                <div className="circle-wrapper">{ratingCircle}</div>
                <div className="btn-wrapper heart-wrapper"><i className={`${heartClass} fa-heart`} /></div>
                <div className="btn-wrapper bookmark-wrapper"><i className="far fa-bookmark"></i></div>

              </div>
            </div>
          </article>
          </div>
        </div>
        <CastContainer cast={cast}/>
      </div>
    );
  }
}

export default MovieDetails;

MovieDetails.propTypes = {
  ovieDetails: PropTypes.object,
  people: PropTypes.object
};
