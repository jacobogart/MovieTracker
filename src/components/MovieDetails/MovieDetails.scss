.MovieDetails-backdrop {
  background-size: cover;
}

.backdrop-cover {
  background: rgba($theme, .95);
  height: 590px;
  display: flex;
  align-items: center;
  justify-content: space-around;
  padding: 10px;
}

.MovieDetails {
  background: transparent;
  color: $off-white;
  display: flex;
  text-align: left;
  width: 70vw;
  height: 60vh;

  .card-img {
    height: 60vh;
    object-fit: cover;
  }

  .card-info {
    margin: 0 20px;
    padding: 0 20px;
    position: relative;
    width: 80vw;

    .card-header {
      // display: flex;
      // align-items: center;

      .card-btns {
        display: flex;
        align-items: center;
        margin: 20px 0;
      }

      .rating-circle {
        border-radius: 50%;
        width: 70px;
        height: 70px;
      }
    
      .circle-bg {
        fill: none;
        stroke: #999;
        stroke-width: 4px;
      }

      .circle {
        fill: none;
        stroke: $accent;
        stroke-width: 4px;
        stroke-linecap: round;
        animation: progress 1s ease-out forwards;
      }

      .rating-num {
        font-size: .8em;
      }

      .btn-wrapper {
        background: transparent;
        border: 3px solid white;
        border-radius: 50%;
        height: 50px;
        margin: 0 10px 0 20px;
        width: 50px;

        &:last-child {
          margin: 0 10px;
        }
      }

      .fa-heart,
      .fa-bookmark {
        color: white;
        font-size: 1.5em;
        position: relative;
        top: 14px;
        left: 13px;
      }

      .fa-heart:hover {
        color: red;
      }

      .fa-bookmark:hover {
        color: $accent;
      }

      .fa-bookmark {
        left: 16px;
      }

      .card-title-wrap {
        display: flex;

        .card-title {
          font-size: 2.5em;
        }

        .card-year {
          font-size: 1.5em;
          margin: 0 0 0 10px;
          position: relative;
          top: 12px;
        }
      }
    }

    .card-overview {
      overflow: auto;
      margin: 30px 0;
      font-size: 1.3em;
    }

    h3 {
      font-size: 1.5em;
    }

    .crew-info {
      display: flex;
      flex-wrap: wrap;

      .crew-member {
        font-size: 1.1em;
        margin: 25px 25px 0 0;
      }
    }
  }
}

@keyframes progress {
  0% {
    stroke-dasharray: 0 100;
  }
}