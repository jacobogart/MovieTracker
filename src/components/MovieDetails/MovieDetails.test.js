import React from "react";
import { shallow } from "enzyme";
import { MovieDetails } from "./MovieDetails";
import { cleanMovie, cleanPeople } from "../../api/__tests__/mockData";

describe("MovieDetails", () => {
  it("should match the snapshot", () => {
    let wrapper = shallow(
      <MovieDetails 
        movieDetails={cleanMovie}
        people={cleanPeople}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});