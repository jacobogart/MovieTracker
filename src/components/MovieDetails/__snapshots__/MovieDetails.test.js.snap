// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails should match the snapshot 1`] = `
<div>
  <article
    className="MovieDetails"
  >
    <img
      alt="Avengers: Endgame poster"
      className="card-img"
      src="https://image.tmdb.org/t/p/w500/or06FN3Dka5tukK1e9sl16pB3iy.jpg"
    />
    <div
      className="card-info"
    >
      <div
        className="card-header"
      >
        <div
          className="circle-wrapper"
        >
          <svg
            className="rating-circle"
            viewBox="0 0 36 36"
          >
            <path
              className="circle-bg"
              d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
            />
            <path
              className="circle"
              d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
              strokeDasharray="86, 100"
            />
            <text
              className="rating-num"
              x="9"
              y="23"
            >
              8.6
            </text>
          </svg>
        </div>
        <div
          className="card-title-wrap"
        >
          <h2
            className="card-title"
          >
            Avengers: Endgame
          </h2>
          <p
            className="card-year"
          >
            (
            2019
            )
          </p>
        </div>
      </div>
      <p
        className="card-overview"
      >
        After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in order to undo Thanos' actions and restore order to the universe once and for all, no matter what consequences may be in store.
      </p>
    </div>
  </article>
  <CastContainer
    cast={
      Array [
        Object {
          "character": "Captain America",
          "headshot": "https://image.tmdb.org/t/p/w500/fdsfdsfsdfa.png",
          "id": 1234,
          "name": "Chris Evans",
        },
        Object {
          "character": "Bucky",
          "headshot": "https://image.tmdb.org/t/p/w500/yeeeeeeet.png",
          "id": 4321,
          "name": "Sebastian Stan",
        },
      ]
    }
  />
</div>
`;
